service_storagefacility/app/migration/EventMigrator.scala

```diff
@@ -0,0 +1,249 @@
+               moveObjEvents.map(_._1) -> mr
+             }
+        tr <- if (theRestEvents.isEmpty)
+               Future.successful(Seq.empty -> MigrationResult())
+             else
+               Future.sequence {
+                 theRestEvents.map {
+                   case (oldId: EventId, (mid: MuseumId, e: MoveEvent)) =>
+                     movDao.insert(mid, e).map((oldId, _))
+
+                   case (oldId: EventId, (mid: MuseumId, e: EnvRequirement)) =>
+                     envDao.insert(mid, e).map((oldId, _))
+
+                   case (oldId: EventId, (mid: MuseumId, e: Observation)) =>
+                     obsDao.insert(mid, e).map((oldId, _))
+
+                   case (oldId: EventId, (mid: MuseumId, e: Control)) =>
+                     ctlDao.insert(mid, e).map((oldId, _))
+                 }
+               }.map { results =>
+                 val r = results.foldLeft((List.empty[EventId], MigrationResult())) {
+                   case (acc, curr) =>
+                     curr match {
+                       case (oldId, MusitSuccess(eid)) =>
+                         val oldIds = acc._1 ::: oldId :: Nil
+                         oldIds -> acc._2.copy(success = acc._2.success + 1)
+
+                       case (oldId, error) =>
+                         logger.warn(s"Something went wrong migrating $oldId")
+                         acc._1 -> acc._2.copy(errors = acc._2.errors :+ oldId)
+                     }
+                 }
+
+                 logger.info(s"Wrote ${r._2.success} of ${theRestEvents.size} events")
+                 r
+               }
+        _ <- migrated(mo._1 ++ tr._1)
+      } yield {
+        val s = mo._2.success + tr._2.success
+        val e = mo._2.errors ++ tr._2.errors
+        MigrationResult(s, e)
+      }
+    }
+  }
+
+  def migrateAll(): Future[Int] = {
+    val startTime: FiniteDuration = System.currentTimeMillis() milliseconds
+
+    logger.warn("Starting data migration of old events")
+
+    val stream = lastWritten
+      .map(l => migrationDao.streamAllEventsFrom(l))
+      .getOrElse(migrationDao.streamAllEvents)
+
+    Source
+      .fromPublisher(stream)
+      .via(new EventConcat)
+      .map { rows =>
+        if (rows.size > 1) {
+          rows.reduceLeft { (acc, curr) =>
+            logger.debug(s"merging acc (${acc._1.id}) with curr (${curr._1.id})")
+            (migrationDao.enrichRelations(acc._1, curr), acc._2, acc._3)
+          }
+        } else {
+          rows.head
+        }
+      }
+      .grouped(1000)
+      .runFoldAsync(MigrationResult()) { (acc, curr) =>
+        convertAndInsert(curr).map { res =>
+          acc.copy(
+            success = acc.success + res.success,
+            errors = acc.errors ++ res.errors
+          )
+        }
+      }
+      .map { res =>
+        val endTime   = System.currentTimeMillis() milliseconds
+        val totalTime = endTime - startTime
```

Comment:
cool...didn't know about that. Not too worried about that for this piece of code though. It'll be deleted in a release or two :-)